examples/load-reporting-service/server/lrs_server.go

```diff
@@ -0,0 +1,161 @@
+	OnStreamClosed(int64)
+	// OnStreamRequest is called once a request is received on a stream.
+	// Returning an error will end processing and close the stream. OnStreamClosed will still be called.
+	OnStreamRequest(int64, *gcpLoadStats.LoadStatsRequest) error
+	// OnStreamResponse is called immediately prior to sending a response on a stream.
+	OnStreamResponse(int64, *gcpLoadStats.LoadStatsResponse)
+}
+
+// NewLrsServer creates handlers from a config watcher and callbacks.
+func NewServer(ctx context.Context, callbacks Callbacks) Server {
+	return &server{callbacks: callbacks, ctx: ctx, lrsCache: make(map[string]map[string]NodeMetadata)}
+}
+
+type server struct {
+	callbacks Callbacks
+	// streamCount for counting bi-di streams
+	streamCount int64
+	ctx         context.Context
+	lrsCache    map[string]map[string]NodeMetadata
+}
+
+type stream interface {
+	grpc.ServerStream
+
+	Send(*gcpLoadStats.LoadStatsResponse) error
+	Recv() (*gcpLoadStats.LoadStatsRequest, error)
+}
+
+// process handles a bi-di stream request
+func (s *server) process(stream stream, reqCh <-chan *gcpLoadStats.LoadStatsRequest) error {
+	// increment stream count
+	streamID := atomic.AddInt64(&s.streamCount, 1)
+
+	if s.callbacks != nil {
+		if err := s.callbacks.OnStreamOpen(stream.Context(), streamID); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Print("Connection ended")
+			return nil
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+
+			if s.callbacks != nil {
+                if err := s.callbacks.OnStreamRequest(streamID, req); err != nil {
+                    return err
+                }
+            }
+
+			clusterName := req.GetNode().GetCluster()
+			nodeId := req.GetNode().GetId()
+			if _, exist := s.lrsCache[clusterName]; !exist {
+				log.Printf("Adding new cluster to cache `%s` with node `%s`", clusterName, nodeId)
+				s.lrsCache[clusterName] = make(map[string]NodeMetadata)
+				s.lrsCache[clusterName][nodeId] = NodeMetadata{
+                    node:   req.GetNode(),
+                    stream: stream,
+                }
+			} else if _, exist := s.lrsCache[clusterName][nodeId]; !exist {
+			    log.Printf("Adding new node `%s` to existing cluster `%s`", nodeId, clusterName)
+			    s.lrsCache[clusterName][nodeId] = NodeMetadata{
+                    node:   req.GetNode(),
+                    stream: stream,
+                }
+			} else{
+				for i := 0; i < len(req.ClusterStats); i++ {
+					if len(req.ClusterStats[i].UpstreamLocalityStats) > 0 {
+						log.Printf("Got stats from cluster `%s` node `%s` - %s", req.Node.Cluster, req.Node.Id, req.GetClusterStats()[i])
+					}
+				}
+			}
+		}
+	}
+}
+
+// handler converts a blocking read call to channels and initiates stream processing
+func (s *server) handler(stream stream) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *gcpLoadStats.LoadStatsRequest)
```

Comment:
this code confuses me a bit. I'm not convinced that we actually need another channel/sentinel to make this work